projects/RabbitMQ.Client/client/impl/WireFormatting.cs

```diff
@@ -700,39 +718,74 @@ public static int WriteShort(Span<byte> span, ushort val)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe int WriteShortstr(Span<byte> span, string val)
+        public static int WriteShortstr(Span<byte> span, string val)
         {
-            int maxLength = span.Length - 1;
-            if (maxLength > byte.MaxValue)
-            {
-                maxLength = byte.MaxValue;
-            }
-            fixed (char* chars = val)
-            fixed (byte* bytes = &span.Slice(1).GetPinnableReference())
+            int bytesWritten = 0;
+            if (!string.IsNullOrEmpty(val))
             {
+                int maxLength = span.Length - 1;
+                if (maxLength > byte.MaxValue)
+                {
+                    maxLength = byte.MaxValue;
+                }
+#if NETCOREAPP
                 try
                 {
-                    int bytesWritten = Encoding.UTF8.GetBytes(chars, val.Length, bytes, maxLength);
-                    span[0] = (byte)bytesWritten;
-                    return bytesWritten + 1;
+                    bytesWritten = UTF8.GetBytes(val, span.Slice(1, maxLength));
                 }
                 catch (ArgumentException)
                 {
                     return ThrowArgumentOutOfRangeException(val, maxLength);
                 }
+#else
+                unsafe
+                {
+                    fixed (char* chars = val)
+                    {
+                        try
+                        {
+                            fixed (byte* bytes = span.Slice(1))
+                            {
+                                bytesWritten = UTF8.GetBytes(chars, val.Length, bytes, maxLength);
+                            }
+                        }
+                        catch (ArgumentException)
+                        {
+                            return ThrowArgumentOutOfRangeException(val, maxLength);
+                        }
+                    }
+                }
+#endif
             }
+
+            span[0] = (byte)bytesWritten;
+            return bytesWritten + 1;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe int WriteLongstr(Span<byte> span, string val)
+#if NETCOREAPP
+        public static int WriteLongstr(Span<byte> span, ReadOnlySpan<char> val)
+        {
+            int bytesWritten = val.IsEmpty ? 0 : UTF8.GetBytes(val, span.Slice(4));
```

Comment:
If we special cased the WriteLongstr for empty values, shouldn't we also do it for the read? => https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/981/files#diff-24820bc65c94384792cf645221a57c741a49e59275b2b0eb34f09167de064bb3R143